Auto/Translation/Assumptions.lean

```diff
@@ -10,16 +10,55 @@ open Lean
 
 namespace Auto
 
+/-- Describes how this fact came about. -/
+inductive Origin where
+  /-- The `<num>`-th definitional equality associated with def `<name>` -/
+  | defEq : Name → Nat → Origin
+  /-- Lemmas hard-wired into Lean-auto -/
+  | hwLem : Name → Origin
+  /-- Lemma from local context -/
+  | lctxLem : Name → Origin
+  /-- User-provided lemma, with the string being a serialisation of the term -/
+  | userLem : String → Origin
+  /-- Lemma from lemma database -/
+  | lemDbLem : Name → Name → Origin
+  /-- Inhabitation fact from local context -/
+  | lctxInh : Name → Origin
+  /-- Inhabitation instance synthesized for canonicalized type -/
+  | tyCanInh : Origin
+  /-- Inductive type recursor -/
+  | recursor : Name → Origin
+  /-- Rewrite -/
+  | rewrite : Origin
+  /-- Lemma proved by native prover -/
+  | byNative : Name → Origin
+  /-- Catch-all for string-based origins -/
+  | generic : String → Origin
```

Comment:
I've begun to add some structure to the stringly-typed proof trees. The goal is to eventually extract a user-readable lemma name from the `DTr` of assertions in the unsat core.